frostsnap_coordinator/src/backup_run.rs

```diff
@@ -1,43 +1,40 @@
 use crate::persist::{Persist, ToStringWrapper};
 use bdk_chain::rusqlite_impl::migrate_schema;
-use frostsnap_core::{AccessStructureId, AccessStructureRef, DeviceId, Gist, KeyId};
+use frostsnap_core::{AccessStructureId, AccessStructureRef, Gist, KeyId};
 use rusqlite::params;
 use std::collections::BTreeMap;
-use std::time::{SystemTime, UNIX_EPOCH};
 use tracing::{event, Level};
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct DisplayBackupState {
+    pub confirmed: bool,
+    pub legacy_display_confirmed: bool,
+}
+
 #[derive(Default, Debug)]
 pub struct BackupState {
-    // Maps each access_structure_ref to a map of device_id -> timestamp.
-    runs: BTreeMap<AccessStructureRef, BTreeMap<DeviceId, Option<u32>>>,
+    // Maps each access_structure_ref to a map of share_index -> complete boolean.
+    runs: BTreeMap<AccessStructureRef, BTreeMap<u32, bool>>,
 }
 
 impl BackupState {
     fn apply_mutation(&mut self, mutation: &Mutation) -> bool {
         match mutation {
-            Mutation::AddDeviceNeedsBackup {
+            Mutation::AddShareNeedsBackup {
                 access_structure_ref,
-                device_id,
+                share_index,
             } => {
                 let run = self.runs.entry(*access_structure_ref).or_default();
-                run.insert(*device_id, None);
+                run.insert(*share_index, false);
                 true
             }
-            Mutation::MarkDeviceComplete {
+            Mutation::MarkShareComplete {
                 access_structure_ref,
-                device_id,
-                timestamp,
+                share_index,
             } => {
-                if let Some(run) = self.runs.get_mut(access_structure_ref) {
-                    if let Some(entry) = run.get_mut(device_id) {
-                        *entry = Some(*timestamp);
-                        true
-                    } else {
-                        false
-                    }
-                } else {
-                    false
-                }
+                let run = self.runs.entry(*access_structure_ref).or_default();
+                run.insert(*share_index, true);
+                true
```

Comment:
We always have a mutation here since it isn't possible to `MarkShareComplete` twice? Otherwise it would be possible to store redundant mutations.

Was wondering if we should return something like `true != run.insert(*share_index, true)` here.

Reply:
I'm not sure what you mean. The logic marks it as done regardless of whether it was in the false state beforehand. This doesn't have much affect on the UI but is a better model and a bit more robust.

Reply:
My point here was about not needing to `mutations.push(mutation);` in the case where `run` already has `true` for that share index, pushing a redundant mutation onto the log. Repeated `mark_backup_complete`s.

But this is probably prevented by the protocol flow.